src/Containers/OhmmsPETE/OhmmsArray.h

```diff
@@ -16,100 +16,84 @@
 
 /** A D-dimensional Array class based on PETE
  *
- *  Equivalent to blitz::Array<T,D>, pooma::Array<D,T>.
- *  No operators are provided.
- *  \todo use aligned_vector<T>
  */
-#include <vector>
-#include <iostream>
+#include <array>
+#include <type_traits>
+#include "OhmmsVector.h"
 
-template<class T, unsigned D>
-struct Array
+template<class T, unsigned D, typename ALLOC = std::allocator<T>>
+class Array
 {
+public:
   using Type_t      = T;
-  using Container_t = std::vector<T>;
-  using This_t      = Array<T, D>;
-
-  size_t Length[D];
-  Container_t X;
+  using Container_t = qmcplusplus::Vector<T, ALLOC>;
+  using This_t      = Array<T, D, ALLOC>;
 
   //default constructor
-  Array()
-  {
-    for (int i = 0; i < D; i++)
-      Length[i] = 0;
-  }
+  Array() { Length.fill(0); }
 
   //copy constructor
   Array(const Array& rhs)
   {
-    resize(rhs);
+    resize(rhs.shape());
     std::copy(rhs.begin(), rhs.end(), X.begin());
   }
 
-  // specialized for 1-Dim
-  Array(size_t n) { resize(n); }
-
-  // specialized for 2-Dim
-  Array(size_t n, size_t m) { resize(n, m); }
-
-  // specialized for 3-Dim
-  Array(size_t l, size_t m, size_t n) { resize(l, m, n); }
-
-  // specialized for 4-Dim
-  Array(size_t l, size_t m, size_t n, size_t o) { resize(l, m, n, o); }
-
-  // specialized for something like TinyVector
-  template<typename ST1>
-  Array(ST1* dims)
+  template<typename SIZET = size_t, typename = std::is_integral<SIZET>>
+  Array(const std::array<SIZET, D>& dims)
   {
     resize(dims);
   }
 
-  // do nothing Destructor
-  ~Array() {}
+  /// Provides specialized constructors with signature (size_1, ... , size_D) for  array dimension D
+  template<typename... Args>
+  Array(Args... sizes)
+  {
+    resize(sizes...);
+  }
 
   inline unsigned dim() const { return D; }
-  inline size_t* shape() const { return &Length[0]; }
+  inline const std::array<size_t, D>& shape() const { return Length; }
   inline size_t size() const { return X.size(); }
   inline size_t size(int i) const { return Length[i]; }
 
   Container_t& storage() { return X; }
 
-  template<typename TT>
-  void resize(const Array<TT, D>& rhs)
+  /// Resize the container. For performance consideration, previous data may or may not get kept.
+  /// Please avoid relying on previous data after resizing.
+  template<typename SIZET = size_t, typename = std::is_integral<SIZET>>
+  void resize(const std::array<SIZET, D>& dims)
   {
-    X.resize(rhs.size());
-    for (int i = 0; i < D; i++)
-      Length[i] = rhs.Length[i];
+    for (int i = 0; i < dims.size(); i++)
+      Length[i] = dims[i];
+    X.resize(full_size(Length));
```

Comment:
In order to help identify cases were the data might get reused, maybe debug builds should set the container to some value (zero?  NaN?  some arbitrary large value?)  (For a future PR)

Reply:
Good suggestion. I can explore filling with a default value. However, the concern is significant slow down in a debug more. So need to try it out.